src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,378 @@
+                        // DO NOTHING FOR THESE UNKNOWN OPTIONS
+                        offset += PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES;
+                        break;
+                }
+
+                if (offset < preloginPayload.Length)
+                {
+                    option = (PreLoginOptions)preloginPayload[offset++];
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Constructs the Prelogin packet and sends it to the server.
+        /// </summary>
+        /// <param name="context">The Prelogin handler context</param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private static async Task CreatePreloginAndSend(PreloginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            Debug.Assert(context.ConnectionContext.TdsStream != null, "A Tds Stream is expected");
+
+            TdsStream tdsStream = context.ConnectionContext.TdsStream;
+            tdsStream.PacketHeaderType = TdsStreamPacketType.PreLogin;
+
+            // Initialize option offset into payload buffer
+            // 5 bytes for each option (1 byte length, 2 byte offset, 2 byte payload length)
+            int optionsHeaderSize = (int)PreLoginOptions.NUMOPT * 5;
+
+            int offset = optionsHeaderSize + 1;
+
+            // Create the payload buffer with the options header and options payload length, which 
+            // is precalculated.
+            byte[] payload = new byte[optionsHeaderSize + 1 + s_optionsPayloadLength];
+
+            for (int option = (int)PreLoginOptions.VERSION; option < (int)PreLoginOptions.NUMOPT; option++)
+            {
+                int optionHeaderOffset = option * 5;
+                payload[optionHeaderOffset] = (byte)option;
+                payload[optionHeaderOffset + 1] = (byte)((offset & 0xff00) >> 8);
+                payload[optionHeaderOffset + 2] = (byte)(offset & 0x00ff);
+                payload[optionHeaderOffset + 3] = (byte)((s_optionsLength[option] & 0xff00) >> 8);
+                payload[optionHeaderOffset + 4] = (byte)(s_optionsLength[option] & 0x00ff);
+                
+                switch (option)
+                {
+                    case (int)PreLoginOptions.VERSION:
+                        Version systemDataVersion = ADP.GetAssemblyVersion();
+
+                        // Major and minor
+                        payload[offset++] = (byte)(systemDataVersion.Major & 0xff);
+                        payload[offset++] = (byte)(systemDataVersion.Minor & 0xff);
+
+                        BinaryPrimitives.WriteUInt16BigEndian(payload.AsSpan(offset, 2), (ushort)systemDataVersion.Build);
+                        offset += 2;
+                        BinaryPrimitives.WriteUInt16LittleEndian(payload.AsSpan(offset, 2), (ushort)systemDataVersion.Revision);
```

Comment:
```suggestion
                        BinaryPrimitives.WriteUInt16BigEndian(payload.AsSpan(offset, 2), (ushort)systemDataVersion.Revision);
```

Reply:
The code which is ported over. 

```
// Sub-build (Little Endian)
payload[payloadLength++] = (byte)(systemDataVersion.Revision & 0xff);
payload[payloadLength++] = (byte)((systemDataVersion.Revision & 0xff00) >> 8);
```

Its inconsistent, but it is what it is.

Reply:
You're right, sorry about that. The TDS spec says that the four-byte version is big-endian and the latter two bytes are little-endian. It then writes every version component in reverse order (build.minor.major) and specifies that the bytes of the only field (build) where endianness would make a difference should be swapped.

I took a packet capture of a pre-login message to help understand the strange behaviour and saw a version of 3.15.23292 with a sub-build of 1 written out as [03, 0f, 5a, fc, 01, 00], so your code's absolutely fine. I just didn't parse the TDS spec properly!

I've also just checked the TRACEID option value; your existing code there is correct too, so I've deleted my comment.

Reply:
Thanks for the verification @edwardneal